### PR TITLE
cost-tracker-stats: add transaction cost distribution stats

### LINKED ISSUE #463: CostTrackerStats: CU distribution stats
#### Problem
- It can be useful to know the distribution of CUs for committed transactions

#### Proposed Solution
- Report min, mean, max, p90 for **actual** cus in "cost-tracker-stats"
- Could accomplish this by adding a histo struct that gets reset on slot boundary and adding transactions here: https://github.com/anza-xyz/agave/blob/master/core/src/banking_stage/qos_service.rs#L198

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
#463 

#### Summary of Changes
- Collect histogram on actual (committed) transaction costs
- Report (min, mean, max, p90) of transaction costs

Fixes #463

### CODE REVIEW

File: core/src/banking_stage/consumer.rs [hunk 1]
```diff
@@ -519,10 +519,16 @@ impl Consumer {
         // once feature `apply_cost_tracker_during_replay` is activated, leader shall no longer
         // adjust block with executed cost (a behavior more inline with bankless leader), it
         // should use requested, or default `compute_unit_limit` as transaction's execution cost.
-        if !bank
+        if bank
```
Comment (apfitzge) [inline core/src/banking_stage/consumer.rs:522]: flipped the if here to be positive instead of negative, since we want to record the unadusted costs into our histogram no.

File: cost-model/src/cost_tracker.rs [hunk 2]
```diff
@@ -83,10 +85,18 @@ impl Default for CostTracker {
...
             ed25519_instruction_signature_count: 0,
+            transaction_cost_histogram: Histogram::new(),
         }
     }
 }
 
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl solana_frozen_abi::abi_example::AbiExample for CostTracker {
+    fn example() -> Self {
+        Self::default()
```
Comment (apfitzge) [inline cost-model/src/cost_tracker.rs:101]: this field on bank is not serialized or deserialized; we don't need to care about a stable abi for it!

File: core/src/banking_stage/qos_service.rs [hunk 3]
```diff
@@ -146,6 +146,31 @@ impl QosService {
         }
     }
 
+    /// Record unadjusted costs for committed transactions.
+    /// This should be called instead of `update_costs` when the feature is
+    /// activated to no longer update costs.
+    pub fn record_unadjusted_costs<'a>(
+        transaction_cost_results: impl Iterator<Item = &'a transaction::Result<TransactionCost>>,
+        transaction_committed_status: Option<&Vec<CommitTransactionDetails>>,
```
Comment (ryoqun) [inline core/src/banking_stage/qos_service.rs:154]: nit: naming `transaction_committed_statuses` or `commit_transaction_details_vec`?

File: core/src/banking_stage/qos_service.rs [hunk 4]
```diff
@@ -146,6 +146,31 @@ impl QosService {
...
+        bank: &Bank,
+    ) {
+        if let Some(transaction_committed_status) = transaction_committed_status {
+            let mut cost_tracker = bank.write_cost_tracker().unwrap();
```
Comment (ryoqun) [inline core/src/banking_stage/qos_service.rs:158]: nit: avoid write lock when transaction_cost_results or transaction_committed_status is empty?

File: cost-model/src/cost_tracker.rs [hunk 5]
```diff
@@ -149,6 +159,18 @@ impl CostTracker {
                 );
             }
         }
+
+        // Add actual cost of transaction to histogram
+        let transaction_cost = estimated_tx_cost
+            .sum()
+            .saturating_sub(estimated_tx_cost.programs_execution_cost())
```
Comment (ryoqun) [inline cost-model/src/cost_tracker.rs:166]: nit: the already-declared `estimated_execution_units` variable can be used.

File: cost-model/src/cost_tracker.rs [hunk 6]
```diff
@@ -149,6 +159,18 @@ impl CostTracker {
...
+            .saturating_add(actual_execution_units);
+        let _ = self.transaction_cost_histogram.increment(transaction_cost);
+    }
+
+    pub fn record_unadjusted_cost(&mut self, tx_cost: &TransactionCost) {
+        let transaction_cost = tx_cost.sum();
+        let _ = self.transaction_cost_histogram.increment(transaction_cost);
     }
```
Comment (ryoqun) [inline cost-model/src/cost_tracker.rs:174]: nit: how about arranging these like this to follow logical hierarchy?:

```rust
    fn update_execution_cost() {
        ...
        let adjusted_cost = estimated_tx_cost.sum()...;
        self.update_committed_cost_histogram(adjusted_cost);
    }

    pub fn update_committed_cost_histogram(&mut self, actual_cost: u64) {
        let _ = self.committed_cost_histogram.increment(actual_cost);
    }

...
// at call site
    pub fn update_committed_cost_histogram<'a>(...) {
        ...
        cost_tracker.update_committed_cost_histogram(tx_cost.sum());
        ...
    }

```

File: core/src/banking_stage/consumer.rs [hunk 7]
```diff
@@ -519,10 +519,16 @@ impl Consumer {
...
             .feature_set
             .is_active(&feature_set::apply_cost_tracker_during_replay::id())
         {
+            QosService::record_unadjusted_costs(
+                transaction_qos_cost_results.iter(),
+                commit_transactions_result.as_ref().ok(),
+                bank,
+            );
+        } else {
```
Comment (ryoqun) [inline core/src/banking_stage/consumer.rs:531]: nit: ... or more like this (alternative to https://github.com/anza-xyz/agave/pull/472#discussion_r1549050656)?:

```rust
// here
QosService::update_costs(
    transaction_qos_cost_results.iter(),
    commit_transactions_result.as_ref().ok(),
    !bank.feature_set.is_active(&feature_set::apply_cost_tracker_during_replay::id()),
);

...
// at definition of the called method (via `update_costs`)
    fn update_committed_transaction_costs<'a>(
        transaction_cost_results: impl Iterator<Item = &'a transaction::Result<TransactionCost>>,
        transaction_committed_status: &Vec<CommitTransactionDetails>,
        bank: &Bank,
        adjust_with_actual_cu: bool,
    ) {
        ...
        let committed_transaction_cost = if adjust_with_actual_cu {
            cost_tracker.update_execution_cost(tx_cost, *compute_units);
            // Adjust with actual executed compute_units
            tx_cost.sum()...
        } else {
            tx_cost.sum()
        }
        cost_tracker.update_committed_cost_histogram(tx_cost.sum());
        ...
    }

...

// lastly, revert all changes made to `update_execution_cost`

```

File: cost-model/src/cost_tracker.rs [hunk 8]
```diff
@@ -88,10 +90,18 @@ impl Default for CostTracker {
             secp256k1_instruction_signature_count: 0,
             ed25519_instruction_signature_count: 0,
             in_flight_transaction_count: 0,
+            transaction_cost_histogram: Histogram::new(),
```
Comment (ryoqun) [inline cost-model/src/cost_tracker.rs:93]: nit: rename to `committed_cost_histogram`? i think `transaction_` here could be a bit redundant. and this histogram is only for committed tasks.

**Review by ryoqun — APPROVED** [review]
lgtm with nits

left somewhat opinionated code arrangement suggestions. they are nits, but i thought current control flow is a bit hard to follow. maybe the primary reason is that both code pathes are actually modified even though we're introducing branching with an existing feature gate. Also, it's a bit hard to tell only histgram is updated after feature activation from method names

**Comment by ryoqun** [comment]
oh, i just noticed that once `apply_cost_tracker_during_replay` is activated, the `"cost_tracker_stats"` metrics will be started to be reported for all of replayed blocks too, in addition to leaders by the banking stage. And, in that case, the new metrics fields will all be zeros, because that code path never update the histogram. So, I think the following code should be updated as well?

https://github.com/anza-xyz/agave/blob/0168e0ab69a631d5998507924debaba1caa566ba/ledger/src/blockstore_processor.rs#L440-L450

Or, prevent reporting this metrics from replaying? That might be noise for the usecase of the metrics, or the high volume nature of its metrics could be concern by itself ..: N validators * M slots

Moreover, i wonder the timeline of the activation of the feature gate by itself... (off-topic: I'm aware unified scheduler code-path need to be updated for the feature gate)

File: cost-model/Cargo.toml [hunk 9]
```diff
@@ -10,14 +10,14 @@ license = { workspace = true }
 edition = { workspace = true }
 
 [dependencies]
+histogram = { workspace = true }
```
Comment (ryoqun) [inline cost-model/Cargo.toml:13]: our `histgram` version is quite old: 6 years old...

https://github.com/anza-xyz/agave/blob/0168e0ab69a631d5998507924debaba1caa566ba/Cargo.toml#L218

https://crates.io/crates/histogram/0.6.9

seems `histgram` got rewrote with new impl recently, along with incompat api:  https://crates.io/crates/histogram/versions, its new design doc: https://observablehq.com/@iopsystems/h2-histogram

so, this is certainly out of scope of this pr, but we need to update the crate eventually...

**Review by ryoqun — COMMENTED** [review]
due to https://github.com/anza-xyz/agave/pull/472#issuecomment-2034607582, I have to cancel my previous r+. sorry about that. Anyway, i think this is not that big deal. i just want to hear about your explicit thinking about it.

File: cost-model/src/cost_tracker.rs (hunk 8 above)
Comment (bw-solana) [inline cost-model/src/cost_tracker.rs:93]: I think it would be nice to track both estimated and actual costs in separate histos
